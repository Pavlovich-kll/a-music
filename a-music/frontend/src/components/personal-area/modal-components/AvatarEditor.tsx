import React from 'react';
import AddAPhotoIcon from '@material-ui/icons/AddAPhoto';
import { Formik, Form } from 'formik';
import Button from '@material-ui/core/Button';
import { toast } from 'react-toastify';
import HTTP from '../../../common/api';

const AVATAR_UPLOAD_ENDPOINT: string = '/user-service/images';
const USER_ID_LOCALSTORAGE: string = 'USER_ID_KEY';
const id = localStorage.getItem(USER_ID_LOCALSTORAGE);
const USER_UPDATE_ENDPOINT: string = `/user-service/users/update/${id}`;
const objToString = (obj: any): string => obj.fileName;

interface IFormikValues {
  avatar: any;
}

const InitialValues: IFormikValues = {
  avatar: null,
};
interface IOnClose {
  onClose: () => any;
}

const AvatarEditor = ({ onClose }: IOnClose) => {
  const [avatar, setAvatar] = React.useState();

  const handleSubmit = async () => {
    if (avatar) {
      const formdata = new FormData();
      formdata.append('file', avatar.content, 'image.jpg');
      const uploadRequest = await HTTP.post(AVATAR_UPLOAD_ENDPOINT, formdata); //ответ верный, fileName: искомое имя
      const userUpdateRequest = await HTTP.patch(USER_UPDATE_ENDPOINT, { avatar: objToString(uploadRequest) });
    }
    if (!avatar) {
      toast.error('You forgot to upload an avatar');
    } else {
      onClose();
    }
  };
  const handleChange = (event: any) => {
    const reader = new FileReader();
    const [content] = event.target.files;
    reader.onloadend = () => {
      setAvatar({ file: reader.result, content });
    };
    reader.readAsDataURL(content);
  };

  return (
    <div className="changeAvatarBlock">
      <Formik initialValues={InitialValues} onSubmit={handleSubmit}>
        <Form>
          <div className="ChangeAvatarBlock">
            <label className="AvatarLabel">
              {avatar ? (
                <img src={avatar.file} />
              ) : (
                <>
                  <AddAPhotoIcon />
                  <span className="title">Add file</span>
                </>
              )}
              <input
                name="file"
                accept="image/*"
                className="input"
                id="contained-button-avatar"
                multiple
                type="file"
                onChange={handleChange}
              />
            </label>
            <div className="status-input"></div>
            <Button className="uploadAvatar" type="submit" variant="contained" color="primary">
              Upload Avatar
            </Button>
          </div>
        </Form>
      </Formik>
    </div>
  );
};

export default AvatarEditor;
