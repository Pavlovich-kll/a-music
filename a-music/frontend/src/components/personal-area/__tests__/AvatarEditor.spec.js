import React from "react";
import AvatarEditor from "../modal-components/AvatarEditor";
import Enzyme from "enzyme";
import { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import {wait} from "@testing-library/react";

Enzyme.configure({ adapter: new Adapter() });

describe('tests AvatarEditor component', () => {
    let wrapper;
    let mockOnClose = jest.fn();

    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState')
     useStateSpy.mockImplementation(() => ['someAvatar.jpeg', setState]);

    beforeEach(()=> {
        wrapper = mount(<AvatarEditor onClose={mockOnClose}/>)
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it("render correctly component", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should after add picture avatar',  () => {
        expect(wrapper.find('.status-input').text()).toEqual('someAvatar.jpeg');
    });

    it('should set state',  async () => {

        wrapper.find('input').simulate('change', {
            target: {name: 'avatar', value:'someAvatar.jpeg'}
        });
        wrapper.find("form").simulate("submit");
        wrapper.update();
        await wait(() => {
            expect(setState).toHaveBeenCalledWith('someAvatar.jpeg');
        });
    });

    it('should click submit button',   async () => {

        wrapper.find('input').simulate('change', {
            target: {name: 'avatar', value:'someAvatar.jpeg'}
        });
        wrapper.find("form").simulate("submit");
        wrapper.update();

        await wait(() => {
            expect(mockOnClose).toBeCalledTimes(1)
        });
    });

});