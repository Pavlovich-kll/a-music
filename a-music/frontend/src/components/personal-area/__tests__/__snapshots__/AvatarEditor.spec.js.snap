// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests AvatarEditor component render correctly component 1`] = `
<AvatarEditor
  onClose={[MockFunction]}
>
  <div
    className="changeAvatarBlock"
  >
    <Formik
      initialValues={
        Object {
          "avatar": null,
        }
      }
      onSubmit={[Function]}
    >
      <Form>
        <form
          action="#"
          onReset={[Function]}
          onSubmit={[Function]}
        >
          <div
            className="ChangeAvatarBlock"
          >
            <label
              className="AvatarLabel"
            >
              <ForwardRef>
                <WithStyles(ForwardRef(SvgIcon))>
                  <ForwardRef(SvgIcon)
                    classes={
                      Object {
                        "colorAction": "MuiSvgIcon-colorAction",
                        "colorDisabled": "MuiSvgIcon-colorDisabled",
                        "colorError": "MuiSvgIcon-colorError",
                        "colorPrimary": "MuiSvgIcon-colorPrimary",
                        "colorSecondary": "MuiSvgIcon-colorSecondary",
                        "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                        "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                        "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                        "root": "MuiSvgIcon-root",
                      }
                    }
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M3 4V1h2v3h3v2H5v3H3V6H0V4h3zm3 6V7h3V4h7l1.83 2H21c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H5c-1.1 0-2-.9-2-2V10h3zm7 9c2.76 0 5-2.24 5-5s-2.24-5-5-5-5 2.24-5 5 2.24 5 5 5zm-3.2-5c0 1.77 1.43 3.2 3.2 3.2s3.2-1.43 3.2-3.2-1.43-3.2-3.2-3.2-3.2 1.43-3.2 3.2z"
                      />
                    </svg>
                  </ForwardRef(SvgIcon)>
                </WithStyles(ForwardRef(SvgIcon))>
              </ForwardRef>
              <span
                className="title"
              >
                Add file
              </span>
              <input
                accept="image/*"
                className="input"
                id="contained-button-avatar"
                multiple={true}
                name="avatar"
                onChange={[Function]}
                type="file"
              />
            </label>
            <div
              className="status-input"
            >
              someAvatar.jpeg
            </div>
            <WithStyles(ForwardRef(Button))
              className="uploadAvatar"
              color="primary"
              type="submit"
              variant="contained"
            >
              <ForwardRef(Button)
                className="uploadAvatar"
                classes={
                  Object {
                    "colorInherit": "MuiButton-colorInherit",
                    "contained": "MuiButton-contained",
                    "containedPrimary": "MuiButton-containedPrimary",
                    "containedSecondary": "MuiButton-containedSecondary",
                    "containedSizeLarge": "MuiButton-containedSizeLarge",
                    "containedSizeSmall": "MuiButton-containedSizeSmall",
                    "disableElevation": "MuiButton-disableElevation",
                    "disabled": "Mui-disabled",
                    "endIcon": "MuiButton-endIcon",
                    "focusVisible": "Mui-focusVisible",
                    "fullWidth": "MuiButton-fullWidth",
                    "iconSizeLarge": "MuiButton-iconSizeLarge",
                    "iconSizeMedium": "MuiButton-iconSizeMedium",
                    "iconSizeSmall": "MuiButton-iconSizeSmall",
                    "label": "MuiButton-label",
                    "outlined": "MuiButton-outlined",
                    "outlinedPrimary": "MuiButton-outlinedPrimary",
                    "outlinedSecondary": "MuiButton-outlinedSecondary",
                    "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
                    "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
                    "root": "MuiButton-root",
                    "sizeLarge": "MuiButton-sizeLarge",
                    "sizeSmall": "MuiButton-sizeSmall",
                    "startIcon": "MuiButton-startIcon",
                    "text": "MuiButton-text",
                    "textPrimary": "MuiButton-textPrimary",
                    "textSecondary": "MuiButton-textSecondary",
                    "textSizeLarge": "MuiButton-textSizeLarge",
                    "textSizeSmall": "MuiButton-textSizeSmall",
                  }
                }
                color="primary"
                type="submit"
                variant="contained"
              >
                <WithStyles(ForwardRef(ButtonBase))
                  className="MuiButton-root MuiButton-contained uploadAvatar MuiButton-containedPrimary"
                  component="button"
                  disabled={false}
                  focusRipple={true}
                  focusVisibleClassName="Mui-focusVisible"
                  type="submit"
                >
                  <ForwardRef(ButtonBase)
                    className="MuiButton-root MuiButton-contained uploadAvatar MuiButton-containedPrimary"
                    classes={
                      Object {
                        "disabled": "Mui-disabled",
                        "focusVisible": "Mui-focusVisible",
                        "root": "MuiButtonBase-root",
                      }
                    }
                    component="button"
                    disabled={false}
                    focusRipple={true}
                    focusVisibleClassName="Mui-focusVisible"
                    type="submit"
                  >
                    <button
                      className="MuiButtonBase-root MuiButton-root MuiButton-contained uploadAvatar MuiButton-containedPrimary"
                      disabled={false}
                      onBlur={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={0}
                      type="submit"
                    >
                      <span
                        className="MuiButton-label"
                      >
                        Upload Avatar
                      </span>
                      <WithStyles(memo)
                        center={false}
                      >
                        <ForwardRef(TouchRipple)
                          center={false}
                          classes={
                            Object {
                              "child": "MuiTouchRipple-child",
                              "childLeaving": "MuiTouchRipple-childLeaving",
                              "childPulsate": "MuiTouchRipple-childPulsate",
                              "ripple": "MuiTouchRipple-ripple",
                              "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                              "rippleVisible": "MuiTouchRipple-rippleVisible",
                              "root": "MuiTouchRipple-root",
                            }
                          }
                        >
                          <span
                            className="MuiTouchRipple-root"
                          >
                            <TransitionGroup
                              childFactory={[Function]}
                              component={null}
                              exit={true}
                            />
                          </span>
                        </ForwardRef(TouchRipple)>
                      </WithStyles(memo)>
                    </button>
                  </ForwardRef(ButtonBase)>
                </WithStyles(ForwardRef(ButtonBase))>
              </ForwardRef(Button)>
            </WithStyles(ForwardRef(Button))>
          </div>
        </form>
      </Form>
    </Formik>
  </div>
</AvatarEditor>
`;
